### PR TITLE
Add a ScrollController parameter to NestedScrollView

### LINKED ISSUE #10925: NestedScrollView throws exception caused by type mismatch
I tried to scroll NestedScrollView using by ScrollController.animateTo, then I got this exception:

```
══╡ EXCEPTION CAUGHT BY GESTURE ╞══
The following assertion was thrown while handling a gesture:
type '_Future<List>' is not a subtype of type 'Future<Null>' of 'function result' where
  _Future is from dart:async
  List is from dart:core
  Future is from dart:async
  Null is from dart:core
Either the assertion indicates an error in the framework itself, or we should provide substantially
more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new
When the exception was thrown, this was the stack:
#0      _NestedScrollCoordinator.animateTo (package:flutter/src/widgets/nested_scroll_view.dart:429:19)
#1      _NestedScrollPosition.animateTo (package:flutter/src/widgets/nested_scroll_view.dart:711:24)
#2      ScrollController.animateTo (package:flutter/src/widgets/scroll_controller.dart:147:37)
#3      scrollToTop (package:myapp/widget/app_bar_title.dart:6:22)
#4      MyTabBar.onTapDown (package:myapp/page/home.dart:260:9)
#5      MyTabBar.build.<anonymous closure> (package:myapp/page/home.dart:244:9)
#6      TapGestureRecognizer._checkDown.<anonymous closure> (package:flutter/src/gestures/tap.dart:138:57)
#7      GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:71:24)
#8      TapGestureRecognizer._checkDown (package:flutter/src/gestures/tap.dart:138:9)
#9      TapGestureRecognizer.didExceedDeadline (package:flutter/src/gestures/tap.dart:110:5)
#11     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:366)
#12     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:394)
#13     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
(elided one frame from package dart:async-patch)
Handler: onTapDown
Recognizer:
  TapGestureRecognizer#f82c9(GestureRecognizerState.possible)
═════════════════════
```

_NestedScrollCoordinator.animateTo's return statement is this:

```dart
return Future.wait<Null>(resultFutures);
```

Future.wait<Null> returns `Future<List<Null>>` but animateTo should return `Future<Null>`;

## Flutter Doctor

```
[✓] Flutter (on Mac OS X 10.12.5 16F73, locale ja-JP, channel master)
    • Flutter at /Applications/flutter
    • Framework revision 0426fb60b9 (3 hours ago), 2017-06-22 17:57:59 -0700
    • Engine revision b0dee695ec
    • Tools Dart version 1.24.0-dev.6.7

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users/najeira/Library/Android/sdk
    • Platform android-25, build-tools 25.0.3
    • ANDROID_HOME = /Users/najeira/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.3.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.3, Build version 8E3004b
    • ios-deploy 1.9.1
    • CocoaPods version 1.2.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Gradle version 3.2
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Ultimate Edition (version 2017.1.4)
    • Flutter plugin version 14.0
    • Dart plugin version 171.4694.29
```

### LINKED ISSUE #11153: `NestedScrollView` should expose a `controller`
It's not currently possible to set the `initialScrollOffset` or read the `offset` or call `jumpTo` / `animateTo` when using `NestedScrollView`, and this is something that would be useful:

- if you want a pinned `AppBar` header to start out in a non-expanded state
- if you want the back button to do something different depending on the scroll position
- if you want to animate to a particular position when an element is tapped

### PR DESCRIPTION
**Pull request by HansMuller**
The NestedScrollView changes were contributed by @collinjackson 

Fixes #10925 
Fixes #11153

### CODE REVIEW

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 1]
```diff
@@ -135,7 +135,7 @@ class _NestedScrollViewState extends State<NestedScrollView> {
     return new CustomScrollView(
       scrollDirection: widget.scrollDirection,
       reverse: widget.reverse,
-      physics: new ClampingScrollPhysics(parent: widget.physics),
+      physics: widget.physics ?? const ClampingScrollPhysics(),
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: can you add a test for this change?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 2]
```diff
@@ -170,12 +170,16 @@ class _NestedScrollMetrics extends FixedScrollMetrics {
 typedef ScrollActivity _NestedScrollActivityGetter(_NestedScrollPosition position);
 
 class _NestedScrollCoordinator implements ScrollActivityDelegate, ScrollHoldController {
-  _NestedScrollCoordinator(this._context, double initialScrollOffset) {
+  _NestedScrollCoordinator(this._context, this._parent) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: why "parent"?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 3]
```diff
@@ -426,7 +430,7 @@ class _NestedScrollCoordinator implements ScrollActivityDelegate, ScrollHoldCont
         return innerActivity;
       },
     );
-    return Future.wait<Null>(resultFutures);
+    await Future.wait<Null>(resultFutures);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: what is the purpose of this change?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 4]
```diff
@@ -170,12 +170,16 @@ class _NestedScrollMetrics extends FixedScrollMetrics {
 typedef ScrollActivity _NestedScrollActivityGetter(_NestedScrollPosition position);
 
 class _NestedScrollCoordinator implements ScrollActivityDelegate, ScrollHoldController {
-  _NestedScrollCoordinator(this._context, double initialScrollOffset) {
+  _NestedScrollCoordinator(this._context, this._parent) {
+    double initialScrollOffset = 0.0;
+    if (_parent != null)
+      initialScrollOffset = _parent.hasClients ? _parent.positions.first.pixels : _parent.initialScrollOffset;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: the "first" here is a bug factory. For example, what if there's a scroll view prior to this one in the layout? it'll be the first scroll position in the default scroll controller, but it has no particular relationship with this scroll view.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 3 above)
Comment (collinjackson) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Future.wait doesn't return null, so Flutter will throw an error in the logs complaining about the type mismatch.

Comment (collinjackson) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: See #10925

Comment (Hixie) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Ah, I see. Is there a test that verifies this particular issue?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 1 above)
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: I've reverted this change. We'll have to revisit it in a future PR.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 3 above)
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: The new tests are tripped up by this.

```
00:01 +3 ~1: - NestedScrollView with a ScrollController
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following assertion was thrown running a test:
type '_Future<List>' is not a subtype of type 'Future<Null>' of 'function result' where
  _Future is from dart:async
  List is from dart:core
  Future is from dart:async
  Null is from dart:core
```

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 2 above)
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: It matches the similarly unqualified "updateParent()" method name.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 3 above)
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Just to be clear: if I reintroduce the error, the new tests catch it.

**Comment by HansMuller** [comment]
I've eliminated the `scrollController.positions.first` hack and restored the `@protected` annotation to the positions method.

I added a TrackingScrollController class, which remembers the last scroll position that changed and reports its `pixels` value as the controller's initialScrollOffset. NestedScrollView now just uses its controller's initialScrollOffset to set up its _NestedScrollControllers.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 3 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart]: great!

File: packages/flutter/lib/src/widgets/scroll_controller.dart [hunk 5]
```diff
@@ -266,3 +268,78 @@ class ScrollController extends ChangeNotifier {
...
+  Map<ScrollPosition, VoidCallback> _positionToListener = <ScrollPosition, VoidCallback>{};
+  ScrollPosition _lastUpdated;
+
+  /// The last [ScrollPosition] to change. Returns null if there aren't any
+  /// attached scroll positions or there hasn't been any scrolling yet.
+  ScrollPosition get mostRecentlyUpdatedPosition => _lastUpdated;
+
+  /// Returns the scroll offset of the [mostRecentlyUpdatedPosition] or 0.0.
+  @override
+  double get initialScrollOffset => _lastUpdated?.pixels ?? 0.0;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/scroll_controller.dart]: Rather than returning 0.0 by default, return super.initialScrollOffset, and then have TrackingScrollController take an initialScrollOffset which it passes to the parent. That way, you can still set an initial scroll offset.

File: packages/flutter/test/widgets/nested_scroll_view_test.dart [hunk 6]
```diff
@@ -6,17 +6,18 @@ import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter_test/flutter_test.dart';
 
-Widget buildTest() {
+Widget buildTest({ ScrollController controller, String title:'TTTTTTTT' }) {
```
Comment (Hixie) [inline packages/flutter/test/widgets/nested_scroll_view_test.dart]: nit: space after colon

**Comment by Hixie** [comment]
LGTM
